File: download.js
const { initializeApp, cert } = require("firebase-admin/app");
const { getFirestore } = require("firebase-admin/firestore");
const fs = require("fs");
const path = require("path");

// Укажите путь к вашему файлу ключа serviceAccountKey.json
const serviceAccountPath = path.join(__dirname, "serviceAccountKey.json");

if (!fs.existsSync(serviceAccountPath)) {
  console.error(`Ошибка: Файл '${serviceAccountPath}' не найден!`);
  console.error("Пожалуйста, скачайте ваш приватный ключ из консоли Firebase (Project Settings -> Service Accounts -> Generate new private key)");
  console.error("и положите его в эту папку под именем 'serviceAccountKey.json'.");
  process.exit(1);
}

const serviceAccount = require(serviceAccountPath);

initializeApp({
  credential: cert(serviceAccount)
});

const db = getFirestore();

async function downloadData() {
  try {
    const dataDir = path.join(__dirname, "data");
    if (!fs.existsSync(dataDir)) {
      fs.mkdirSync(dataDir, { recursive: true });
    }

    // 1. Выгрузка категорий
    console.log("Получение категорий из Firestore...");
    const categoriesSnapshot = await db.collection("categories").get();
    const categories = [];
    categoriesSnapshot.forEach(doc => {
      categories.push(doc.data());
    });

    // Сортировка категорий для стабильности файла (например, по ID)
    categories.sort((a, b) => {
      if (a.id < b.id) return -1;
      if (a.id > b.id) return 1;
      return 0;
    });

    const categoriesPath = path.join(dataDir, "categories.json");
    fs.writeFileSync(categoriesPath, JSON.stringify(categories, null, 2), "utf8");
    console.log(`Успешно сохранено ${categories.length} категорий в '${categoriesPath}'`);

    // 2. Выгрузка товаров
    console.log("Получение товаров из Firestore...");
    const productsSnapshot = await db.collection("products").get();
    const products = [];
    productsSnapshot.forEach(doc => {
      const data = doc.data();
      // Убедимся, что ID числовой, если в файле он был числовым
      if (data.hasOwnProperty("id") && !isNaN(data.id)) {
        data.id = Number(data.id);
      }
      products.push(data);
    });

    // Сортировка товаров по числовому ID
    products.sort((a, b) => a.id - b.id);

    const productsPath = path.join(dataDir, "products.json");
    fs.writeFileSync(productsPath, JSON.stringify(products, null, 2), "utf8");
    console.log(`Успешно сохранено ${products.length} товаров в '${productsPath}'`);

    console.log("Импорт из базы (выгрузка) завершен!");
  } catch (error) {
    console.error("Ошибка при получении данных:", error);
  }
}

downloadData();

File: download.py
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore
import json
import os

# Укажите путь к вашему файлу ключа serviceAccountKey.json
# По умолчанию скрипт ищет его в той же папке
cred_path = os.path.join(os.path.dirname(__file__), "serviceAccountKey.json")

if not os.path.exists(cred_path):
    print(f"Ошибка: Файл '{cred_path}' не найден!")
    print("Пожалуйста, скачайте ваш приватный ключ из консоли Firebase (Project Settings -> Service Accounts -> Generate new private key)")
    print("и положите его в эту папку под именем 'serviceAccountKey.json'.")
    exit(1)

cred = credentials.Certificate(cred_path)
firebase_admin.initialize_app(cred)

db = firestore.client()

def download_data():
    try:
        data_dir = os.path.join(os.path.dirname(__file__), "data")
        if not os.path.exists(data_dir):
            os.makedirs(data_dir, exist_ok=True)

        # 1. Выгрузка категорий
        print("Получение категорий из Firestore...")
        categories_ref = db.collection("categories").stream()
        categories = []
        for doc in categories_ref:
            categories.append(doc.to_dict())

        # Сортировка для стабильности файла
        categories.sort(key=lambda x: x.get("id", ""))

        categories_file = os.path.join(data_dir, "categories.json")
        with open(categories_file, "w", encoding="utf-8") as f:
            json.dump(categories, f, indent=2, ensure_ascii=False)
        print(f"Успешно сохранено {len(categories)} категорий в '{categories_file}'")

        # 2. Выгрузка товаров
        print("Получение товаров из Firestore...")
        products_ref = db.collection("products").stream()
        products = []
        for doc in products_ref:
            data = doc.to_dict()
            if "id" in data:
                try:
                    data["id"] = int(data["id"])
                except (ValueError, TypeError):
                    pass
            products.append(data)

        # Сортировка товаров по числовому ID
        products.sort(key=lambda x: x.get("id", 0))

        products_file = os.path.join(data_dir, "products.json")
        with open(products_file, "w", encoding="utf-8") as f:
            json.dump(products, f, indent=2, ensure_ascii=False)
        print(f"Успешно сохранено {len(products)} товаров в '{products_file}'")

        print("Импорт из базы (выгрузка) завершен!")
    except Exception as e:
        print("Ошибка при получении данных:", e)

if __name__ == "__main__":
    download_data()
